core/templates/domain/topic/TopicObjectFactory.ts

```diff
@@ -541,6 +558,31 @@ export class Topic {
     return this._uncategorizedSkillSummaries.slice();
   }
 
+  getDiagnosticTestSkillSummaries(): ShortSkillSummary[] {
+    return this._skillsForDiagnosticTest.slice();
+  }
+
+  setDiagnosticTestSkillSummaries(
+      skillsForDiagnosticTest: ShortSkillSummary[]): void {
+    this._skillsForDiagnosticTest = skillsForDiagnosticTest;
+  }
+
+  getSkillSummariesThatAreNotInDiagnosticTest(): ShortSkillSummary[] {
+    let topicSkillSummaries = cloneDeep(this._uncategorizedSkillSummaries);
+    let subtopics = this._subtopics;
+    for (let i = 0; i < subtopics.length; i++) {
+      topicSkillSummaries = topicSkillSummaries.concat(
+        subtopics[i].getSkillSummaries());
+    }
+    let diagnosticTestSkillSummaries = this.getDiagnosticTestSkillSummaries();
+
+    return topicSkillSummaries.filter(object1 => {
```

Comment:
This isn't particularly efficient. Instead, build a hashmap mapping each skill ID in the topic to true, and then just lookup the keys from the diagnostic test. Otherwise you're doing a nested for loop here, which is unnecessarily slow.

Reply:
Updated